config/config-api/src/main/java/com/thoughtworks/go/config/ArtifactConfig.java

```diff
@@ -17,111 +17,204 @@
+        return StringUtils.isBlank(destination) ? getDefaultFolder() : FilenameUtils.separatorsToUnix(destination);
     }
 
     public String getSource() {
         return FilenameUtils.separatorsToUnix(source);
     }
 
+    public void setSource(String source) {
+        this.source = StringUtils.trim(source);
+    }
+
     public void setDestination(String destination) {
         this.destination = StringUtils.trim(destination);
     }
 
-    public boolean equals(Object other) {
-        return this == other || other != null && other instanceof ArtifactConfig && equals((ArtifactConfig) other);
+    public ArtifactType getArtifactType() {
+        return artifactType;
+    }
+
+    public String getId() {
+        return id;
     }
 
-    private boolean equals(ArtifactConfig other) {
-        if (destination != null ? !destination.equals(other.destination) : other.destination != null) {
-            return false;
-        }
-        return !(source != null ? !source.equals(other.source) : other.source != null) && artifactType.equals(other.artifactType);
+    public void setId(String id) {
+        this.id = id;
+    }
 
+    public String getStoreId() {
+        return storeId;
     }
 
-    public int hashCode() {
-        int result = 0;
-        result = 31 * result + (source != null ? source.hashCode() : 0);
-        result = 31 * result + (destination != null ? destination.hashCode() : 0);
-        result = 31 * result + (artifactType != null ? artifactType.hashCode() : 0);
-        return result;
+    public void setStoreId(String storeId) {
+        this.storeId = storeId;
     }
 
     public String toString() {
         return MessageFormat.format("Artifact of type {0} copies from {1} to {2}", artifactType, source, destination);
     }
 
-    public ArtifactType getArtifactType() {
-        return artifactType;
+    public String getArtifactTypeValue() {
+        switch (artifactType) {
+            case file:
+                return "Build Artifact";
+            case unit:
+                return "Test Artifact";
+            case plugin:
+                return "Pluggable Artifact";
+        }
+        return null;
     }
 
     protected void setArtifactType(ArtifactType artifactType) {
         this.artifactType = artifactType;
     }
 
-    public String getArtifactTypeValue() {
-        return ARTIFACT_PLAN_DISPLAY_NAME;
-    }
-
-    @Override
     public boolean validateTree(ValidationContext validationContext) {
         validate(validationContext);
         return errors().isEmpty();
     }
 
+    @Override
     public void validate(ValidationContext validationContext) {
-        if (!StringUtils.isBlank(destination) && (!(destination.equals(DEFAULT_ROOT.getPath()) || new FilePathTypeValidator().isPathValid(destination)))) {
+        if (artifactType == ArtifactType.plugin) {
+            validatePluggableArtifact(validationContext);
+        } else {
+            validateBuiltInArtifact(validationContext);
```

Comment:
Add more validations - 
1. If id and store id are given, src and dest must not be given
2. If src and destination are provided, pluggable artifact properties, id and store id must not be given.